core/src/mock_alpenglow_consensus.rs

```diff
@@ -0,0 +1,1024 @@
+                *peer,
+                PeerData {
+                    stake,
+                    address: ag_addr,
+                    relative_toa: [None; NUM_VOTOR_TYPES],
+                },
+            );
+            state.total_staked += stake;
+        }
+        trace!(
+            "Prepared for slot {slot}, total stake is {}",
+            state.total_staked
+        );
+        Ok(())
+    }
+
+    /// Collects votes and changes states
+    fn listener_thread(
+        self_state: Arc<StateArray>,
+        should_exit: Arc<AtomicBool>,
+        my_id: Pubkey,
+        socket: Arc<UdpSocket>,
+        command_sender: Sender<SendCommand>,
+    ) {
+        socket
+            .set_read_timeout(Some(Duration::from_secs(1)))
+            .unwrap();
+        trace!("Listener thread started");
+        // Set aside enough space to fetch multiple packets from the kernel per syscall
+        let mut packets: Vec<Packet> = vec![Packet::default(); 1024];
+        loop {
+            // must wipe all Meta records to reuse the buffer
+            for p in packets.iter_mut() {
+                *p.meta_mut() = Meta::default();
+            }
+
+            if should_exit.load(Ordering::Relaxed) {
+                return;
+            }
+            // recv_mmsg should timeout in 1 second
+            let n = match recv_mmsg(&socket, &mut packets) {
+                // we may have received no packets, in this case we can safely skip the rest
+                Ok(0) => continue,
+                Ok(n) => n,
+                Err(e) => {
+                    match e.kind() {
+                        std::io::ErrorKind::TimedOut | std::io::ErrorKind::WouldBlock => {
+                            0 // no packets received
+                        }
+                        _ => {
+                            error!("Got error {:?} in mock alpenglow RX socket operation, exiting thread", e.raw_os_error());
+                            return;
+                        }
+                    }
+                }
+            };
+
+            for pkt in packets.iter().take(n) {
+                if pkt.meta().size < MOCK_VOTE_HEADER_SIZE {
+                    trace!("Packet too small {}", pkt.meta().size);
+                    continue;
+                }
+                let sender = SocketAddr::new(pkt.meta().addr, pkt.meta().port);
+                let Some(pkt_buf) = pkt.data(..) else {
+                    continue;
+                };
+                let vote_pkt = MockVotePacketHeader::from_bytes(pkt_buf);
+                let pk = Pubkey::new_from_array(vote_pkt.sender);
+                if vote_pkt.signature != SIGNATURE {
+                    trace!("Sigverify failed");
+                    continue;
+                }
+
+                let mut state = get_state_for_slot(&self_state, vote_pkt.slot_number)
+                    .lock()
+                    .unwrap();
+
+                if vote_pkt.slot_number != state.current_slot {
```

Comment:
Will this drop votes for slot N if we get votes for slot N before we've voted on slot N and set the slot in `prepare_to_receive` https://github.com/anza-xyz/agave/blob/e60d9d51348ae8fe2ac5ed68276eaff3d1e929c5/core/src/mock_alpenglow_consensus.rs#L245

Reply:
We have 4 state slots, which get reused once the 4-slot batch is complete. However, if a vote is REALLY late (e.g. 5 slot late) it might land into a slot where we expect a vote for a different slot. This check will discard it.